Feature/update auto welcome

temalo opened the pull request:
# Auto-Welcome System Updates

Implemented robust safeguards to prevent nodes from receiving multiple welcome messages by adding both **in-memory tracking** and **atomic database operations** to eliminate race conditions.

***

## 1. Added In-Memory Tracking Set

**File:** `meshtasticManager.ts:100`

*   Added `welcomingNodes: Set<number>` to track nodes currently being welcomed
*   Prevents race condition when multiple `USER` packets arrive simultaneously before database is updated

***

## 2. Enhanced `checkAutoWelcome` Method

**File:** `meshtasticManager.ts:5835-5960`

*   **Pre-flight check:** Immediately returns if node is already being welcomed *(line 5853)*
*   **Lock mechanism:** Adds node to tracking set before welcoming *(line 5876)*
*   **Try-finally pattern:** Ensures tracking set is cleaned up even if errors occur
*   **Delayed cleanup:** Uses 100ms timeout before removing from tracking set to ensure database write completes *(line 5951)*
*   **Enhanced logging:** Shows ISO timestamp of when node was previously welcomed for better debugging

***

## 3. New Atomic Database Operation

**File:** `database.ts:1806-1815`

*   Added `markNodeAsWelcomedIfNotAlready(nodeNum: number, nodeId: string): boolean` method
*   Uses atomic SQL:
    ```sql
    UPDATE ... WHERE welcomedAt IS NULL
    ```
    to ensure only one process marks the node
*   Returns `true` if node was marked, `false` if already welcomed
*   Prevents duplicate marking even if multiple processes bypass the in-memory check

***

## 4. Updated Tests

### Auto-Welcome Tests

**File:** `meshtasticManager.autowelcome.test.ts:301-351`

*   Added test for successful race-condition-free welcoming
*   Added test verifying behavior when node is already being welcomed by another process
*   Mock setup includes `markNodeAsWelcomedIfNotAlready` method

### Database Migration Tests

**File:** `database.autowelcome-migration.test.ts:361-455`

*   Tests atomic operation marks unwelcomed nodes correctly
*   Tests already-welcomed nodes return false without changes
*   Tests idempotency (calling twice returns different results)
*   Tests nodeId validation requirement
*   Tests non-existent node handling

***

## Race Condition Prevention Strategy

*   **Layer 1:** In-memory `Set` checks if welcome is in progress
*   **Layer 2:** Database check verifies node hasn't been welcomed
*   **Layer 3:** Atomic SQL operation prevents duplicate database writes
*   **Layer 4:** `finally` block ensures cleanup regardless of success/failure

***

## Key Benefits

*   Nodes are guaranteed to receive only **one welcome message**
*   Handles concurrent packet arrivals gracefully
*   No duplicate database entries for `welcomedAt`
*   Better logging for debugging welcome status
*   Maintains backward compatibility with existing migration code

Yeraze commented:
## CI Failure Analysis

The tests are failing due to a **package-lock.json sync issue**, not actual test failures:

```
npm error `npm ci` can only install packages when your package.json and package-lock.json are in sync.
npm error Missing: @types/react@18.3.27 from lock file
npm error Missing: react@18.3.1 from lock file  
npm error Missing: react-dom@18.3.1 from lock file
```

**To fix:** Please merge or rebase from `main` to pick up the latest `package-lock.json` changes, then push again.

```bash
git fetch origin main
git merge origin/main
# resolve any conflicts
git push
```

temalo commented:
I've rebased and pushed and my branch is in sync with main, but the tests are still failing:

C:\DevOps\meshmonitor>git checkout feature/update-auto-welcome
Already on 'feature/update-auto-welcome'
Your branch is up to date with 'origin/feature/update-auto-welcome'.

C:\DevOps\meshmonitor>git rebase upstream main
Successfully rebased and updated refs/heads/main.

C:\DevOps\meshmonitor>git push origin feature/update-auto-welcome --force-with-lease
Everything up-to-date

Yeraze commented:
I'll take a look in the morning and see if I can untangle it

Yeraze commented:
@temalo take a look at #943
